android-instrumentation/src/main/java/co/elastic/apm/android/instrumentation/ui/activities/ActivityLifecyclePlugin.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *	http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.android.instrumentation.ui.activities;
+
+import android.app.Activity;
+
+import androidx.annotation.NonNull;
+
+import net.bytebuddy.build.AndroidDescriptor;
+import net.bytebuddy.description.type.TypeDescription;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import co.elastic.apm.android.instrumentation.ui.common.BaseLifecycleMethodsPlugin;
+
+public class ActivityLifecyclePlugin extends BaseLifecycleMethodsPlugin {
+    private final AndroidDescriptor androidDescriptor;
+
+    public ActivityLifecyclePlugin(AndroidDescriptor androidDescriptor) {
+        this.androidDescriptor = androidDescriptor;
+    }
+
+    @Override
+    public boolean matches(TypeDescription target) {
+        if (androidDescriptor.getTypeScope(target) == AndroidDescriptor.TypeScope.EXTERNAL) {
+            return false;
+        }
+        return !target.getSimpleName().startsWith("Hilt_") && target.isAssignableTo(Activity.class);
+    }
+
+    @NonNull
+    @Override
+    protected Class<?> getAdviceClass() {
+        return ActivityLifecycleMethodAdvice.class;
+    }
+
+    @Override
+    protected Map<String, String> provideOrderedTargetNamesToDescriptors() {
+        Map<String, String> targets = new HashMap<>();
+        targets.put("onCreate", "(Landroid/os/Bundle;)V");
+        targets.put("onStart", "()V");
+        targets.put("onResume", "()V");
+        return targets;
```

Comment:
I've just pushed some changes that would allow to describe methods in a friendlier way by using class types. Since there's a lot of overlap between the Activity and Fragment plugins, I wanted to reuse most of the work in a base class for which their subclasses would only need to provide a list of the methods they want to instrument. Said list is also helpful in order to find out what's the last method available in the target class (in order to close the root span there), so it's helpful to have an ordered list of methods instead of a single matcher, thats why I kept the list but just made it classy :) please let me know if any suggestions on it.

Reply:
Maybe a list of method matchers?

> Said list is also helpful in order to find out what's the last method available in the target class (in order to close the root span there)

Could you elaborate a bit more why the order of the methods in the class matters? IIRC, the order is not necessarily deterministic, especially on Android.

Reply:
Yeah no problem, so take an Activity for example, it has 3 methods called in order when [it starts](https://developer.android.com/guide/components/activities/activity-lifecycle#alc) `onCreate`->`onStart`->`onResume`. Since we're wrapping them all in a single, "root" Span, I need to close this root span in the last method, which would be the `onResume` method. The problem is that, the way Byte Buddy works in Android is in "decorator" mode, so it cannot create new methods but only decorate existing ones, so it is possible that the `onResume` method might not be available in an Activity subclass, and in that case we'd need to close the root span somewhere else, `onStart` would be the next choice, and `onCreate` the last one in case it's the only method available. So currently I'm going through the available methods of our target classes, looking for the last available of those 3, and adding an annotation to it to mark it as the one where the root span needs to be closed. In order to do so, I need to check if the method I'm going through for my target class is a valid target by checking if it's available in a list of "target" methods. I'm doing this check on the class `BaseLifecycleMethodsPlugin` on line 102 if you want to have a look

Reply:
Ah, got it now. Thanks for explaining. I assume we can't instrument the method in the base class (Activity?) as it's part of the provided classes on the device, right?

Reply:
Yeah, unfortunately the Activity is part of the Android SDK 😅 - However, the Fragment isn't (there is a Fragment in the Android SDK, but that's deprecated long ago so we're ignoring it) so we could instrument the Fragment's methods, however, even if we did so, wouldn't there be a risk of closing the root span early? In the use case where our target Fragment subclass does have overridden all the methods, the root span would be closed when our target class calls its `super` method, which is usually the first thing called in an overridden method, so I'd be concerned about cutting off the root span at the beginning of the last method and not covering it fully because of it. Also, I personally think it's a nice side effect the fact that we can only instrument the methods overridden by our customers, because that means that they won't see "extra spans" in their Kibana, they'll only see spans for the methods they are actually using in their classes which I think is nice

Reply:
Makes sense. But are there cases where the user might not have overridden any lifecycle method and we therefore miss instrumenting these activities?

Reply:
It is not impossible, but it's also not realistic that there will be an Activity or Fragment without any of the 3 methods overridden because those are the entry points of each class. But if there happen to be a blank Activity or Fragment that don't do anything when created, then yes, those won't get any spans recorded

Reply:
Alright, thanks. Btw, I'm not trying to imply that this is necessarily a problem, just wanted to understand the trade offs. But it sounds sensible to me.